pkg/resources/management.cattle.io/v3/globalrole/validator.go

```diff
@@ -121,49 +120,42 @@ func (a *admitter) Admit(request *admission.Request) (*admissionv1.AdmissionResp
 		return nil, err
 	}
 
+	// check if the user has the escalate verb
+	hasEscalate, escErr := auth.RequestUserHasVerb(request, gvr, a.sar, escalateVerb, newGR.Name, "")
+	if escErr != nil {
+		logrus.Warnf("Failed to check for the 'escalate' verb on GlobalRoles: %v", escErr)
+	}
+
+	// if we have the escalate verb, no need to check global/cluster/namespaced permissions
+	if hasEscalate {
+		return admission.ResponseAllowed(), nil
+	}
+
 	// check for escalation separately between cluster permissions and global permissions to prevent crossover
 	clusterRules, err := a.grbResolver.GlobalRoleResolver.ClusterRulesFromRole(newGR)
 	if err != nil {
 		return nil, fmt.Errorf("unable to resolve rules for new global role: %w", err)
 	}
-
-	hasEscalate, err := a.isRulesAllowed(request, clusterRules, newGR.Name, a.grbResolver)
+	err = auth.ConfirmNoEscalation(request, clusterRules, "", a.resolver)
 	if err != nil {
 		return admission.ResponseFailedEscalation(err.Error()), nil
 	}
-	if hasEscalate {
-		// if we have the escalate verb, no need to check global permissions
-		return admission.ResponseAllowed(), nil
-	}
 
-	rules := a.grbResolver.GlobalRoleResolver.GlobalRulesFromRole(newGR)
-	// don't need to check if this request was allowed due to escalation since this is the last permission check
-	// if others are added in the future a short-circuit here will be needed like for the clusterRules
-	_, err = a.isRulesAllowed(request, rules, newGR.Name, a.resolver)
+	globalRules := a.grbResolver.GlobalRoleResolver.GlobalRulesFromRole(newGR)
+	err = auth.ConfirmNoEscalation(request, globalRules, "", a.resolver)
 	if err != nil {
 		return admission.ResponseFailedEscalation(err.Error()), nil
 	}
 
-	return admission.ResponseAllowed(), nil
-}
-
-// isRulesAllowed checks if the use of requested rules are allowed by the givenResolver for a given request/user
-// returns an error if the user failed an escalation check, and nil if the request was allowed. Also returns a bool
-// indicating if the allow was due to the user having the escalate verb
-func (a *admitter) isRulesAllowed(request *admission.Request, rules []rbacv1.PolicyRule, grName string, resolver validation.AuthorizationRuleResolver) (bool, error) {
-	err := auth.ConfirmNoEscalation(request, rules, "", resolver)
-	if err != nil {
-		hasEscalate, escErr := auth.RequestUserHasVerb(request, gvr, a.sar, escalateVerb, grName, "")
-		if escErr != nil {
-			logrus.Warnf("Failed to check for the 'escalate' verb on GlobalRoles: %v", escErr)
-			return false, err
-		}
-		if hasEscalate {
-			return true, nil
+	// check for escalation in each namespace of NamespacedRules
+	for namespace, rules := range newGR.NamespacedRules {
+		err = auth.ConfirmNoEscalation(request, rules, namespace, a.resolver)
+		if err != nil {
+			return admission.ResponseFailedEscalation(err.Error()), nil
```

Comment:
To add to this a little bit, another UX that could suck is the following scenario:

1. User tries to create GR with namespaced rules: `nsA: rulesA` & `nsB: rulesB`. Both fails but user only sees nsA that fails first.
2. User adjust rules on `nsA`.
3. User tries again to create GR. Fails with `nsB`.
4. User adjust rules on `nsB`.
5. User tries again to create GR. Fails again on `nsA`.

Again, the fact that the map is not deterministic combined with returning on first error means the user can't know _for sure_ that the adjustments made were correct.

For this reason I'd be in favor of trying all the rules and returning all errors. I think the performance hit would be negligible since that's only done when a `GlobalRole` is either created or updated, which is most probably not done often.

Reply:
You both make some great points here. I think returning a `multierror` with all the escalation failures (in Rules, ClusterRules and NamespacedRules) would definitely be an improvement on the UX. While you could infer that a single escalation failure may be relevant to the other Rules, having a full picture makes a lot of sense.

I think the performance hit is preferable to having partial information.